src/mlpack/core/dists/gamma_distribution.cpp

```diff
@@ -117,3 +135,84 @@ void GammaDistribution::Train(const arma::vec& logMeanxVec,
     beta(row) = meanx / aEst;
   }
 }
+
+// Returns the probability of the provided observations.
+void GammaDistribution::Probability(const arma::mat& observations, 
+                                    arma::vec& probabilities) const
+{
+  size_t numObs = observations.n_cols;
+
+  // Set all equal to 1 (multiplication neutral).
+  probabilities.ones(numObs);
+
+  // Compute denominator only once for each dimension.
+  arma::vec denominators(alpha.n_elem);
+  for (size_t d = 0; d < alpha.n_elem; ++d)
+    denominators(d) = std::tgamma(alpha(d)) * std::pow(beta(d), alpha(d));
+
+  // Compute probability of each observation.
+  for (size_t i = 0; i < numObs; ++i)
+  {
+    for (size_t d = 0; d < observations.n_rows; ++d)
+    {
+      // Compute probability using Multiplication Law.
+      probabilities(i) *= 
+        std::pow(observations(d, i), alpha(d) - 1) *
+        std::exp(-observations(d, i) / beta(d)) /
+        denominators(d);
+    }
+  }
+}
+
+// Returns the probability of one observation (x) for one of the Gamma's
+// dimensions.
+double GammaDistribution::Probability(double x, size_t dim) const
+{
+  return 
+    std::pow(x, alpha(dim) - 1) * std::exp(-x / beta(dim)) / 
+    (std::tgamma(alpha(dim)) * std::pow(beta(dim), alpha(dim)));
+}
+
+// Returns the log probability of the provided observations.
+void GammaDistribution::LogProbability(const arma::mat& observations, 
+                                       arma::vec& LogProbabilities) const
+{
+  size_t numObs = observations.n_cols;
+
+  // Set all equal to 0 (addition neutral).
+  LogProbabilities.zeros(numObs);
+
+  // Compute denominator only once for each dimension.
+  arma::vec denominators(alpha.n_elem);
+  for (size_t d = 0; d < alpha.n_elem; ++d)
+    denominators(d) = std::tgamma(alpha(d)) * std::pow(beta(d), alpha(d));
+
+  // Compute probability of each observation.
+  for (size_t i = 0; i < numObs; ++i)
+  {
+    for (size_t d = 0; d < observations.n_rows; ++d)
+    {
+      // Compute probability using Multiplication Law and Logarithm addition
+      // property.
+      LogProbabilities(i) += std::log( 
+        std::pow(observations(d, i), alpha(d) - 1) 
+        * std::exp(-observations(d, i) / beta(d)) 
+        / denominators(d));
+    }
+  }
+}
+
+// Returns a gamma-random d-dimensional vector.
+arma::vec GammaDistribution::Random() const
+{
+  arma::vec randVec(alpha.n_elem);
+
+  std::default_random_engine generator;
```

Comment:
I think you should use `mlpack::math::randGen` here instead, so that if the user sets the random seed, the results when they call `GammaDistribution::Random()` are deterministic.  (`randGen` is defined in `mlpack/core/math/random.hpp`.)